core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala

```diff
@@ -105,10 +107,20 @@ class SparkHadoopWriter(@transient jobConf: JobConf)
   def commit() {
     val taCtxt = getTaskContext()
     val cmtr = getOutputCommitter()
+    val dagSchedulerActor =
+      AkkaUtils.makeDriverRef("DAGScheduler", SparkEnv.get.conf, SparkEnv.get.actorSystem)
+    val askTimeout = AkkaUtils.askTimeout(SparkEnv.get.conf)
     if (cmtr.needsTaskCommit(taCtxt)) {
       try {
-        cmtr.commitTask(taCtxt)
-        logInfo (taID + ": Committed")
+        val canCommit: Boolean = AkkaUtils.askWithReply(
+          AskPermissionToCommitOutput(jobID, splitID, attemptID), dagSchedulerActor, askTimeout)
```

Comment:
You could send a message from the driver node that launched the speculated task to the executor that was running the original copy of the task. It would be tricky to avoid a second race in this solution however.